dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,320 @@
+        get_list = gdrive_retry(lambda: next(file_list, None))
+
+        # Fetch pages until None is received, lazily flatten the thing
+        return cat(iter(get_list, None))
+
+    def cache_root_dirs(self):
+        cached_dirs = {}
+        cached_ids = {}
+        for dir1 in self.gdrive_list_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            cached_ids[dir1["id"]] = dir1["title"]
+        return cached_dirs, cached_ids
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            with open(
+                self.gdrive_user_credentials_path, "w"
+            ) as credentials_file:
+                credentials_file.write(
+                    os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA)
+                )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        GoogleAuth.DEFAULT_SETTINGS["client_config"] = {
+            "client_id": self.gdrive_client_id,
+            "client_secret": self.gdrive_client_secret,
+            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+            "token_uri": "https://oauth2.googleapis.com/token",
+            "revoke_uri": "https://oauth2.googleapis.com/revoke",
+            "redirect_uri": "",
+        }
+        GoogleAuth.DEFAULT_SETTINGS["save_credentials"] = True
+        GoogleAuth.DEFAULT_SETTINGS["save_credentials_backend"] = "file"
+        GoogleAuth.DEFAULT_SETTINGS[
+            "save_credentials_file"
+        ] = self.gdrive_user_credentials_path
+        GoogleAuth.DEFAULT_SETTINGS["get_refresh_token"] = True
+        GoogleAuth.DEFAULT_SETTINGS["oauth_scope"] = [
+            "https://www.googleapis.com/auth/drive",
+            "https://www.googleapis.com/auth/drive.appdata",
+        ]
+
+        # Pass non existent settings path to force DEFAULT_SETTINGS loading
+        gauth = GoogleAuth(settings_file="")
+        gauth.CommandLineAuth()
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            os.remove(self.gdrive_user_credentials_path)
+
+        gdrive = GoogleDrive(gauth)
+        return gdrive
+
+    def create_drive_item(self, parent_id, title):
+        return gdrive_retry(
+            lambda: self.gdrive_create_folder(title, parent_id)
+        )()
+
+    def get_drive_item(self, name, parents_ids):
+        if not parents_ids:
+            return None
+        query = " or ".join(
+            "'{}' in parents".format(parent_id) for parent_id in parents_ids
+        )
+
+        query += " and trashed=false and title='{}'".format(name)
+
+        item_list = gdrive_retry(lambda: self.gdrive_list_file(query))()
+        return next(iter(item_list), None)
+
+    def resolve_remote_file(self, parents_ids, path_parts, create):
```

Comment:
Sounds good.

> The name of this method and a return item in the end suggests that file and item are the same. Are they? Or file sometimes means item and sometimes path?

`item` was used as common for `file` or `directory` on remote, but PyDrive use just `GoogleDriveFile` for both.
It was supposed that method will able to create `item` of requested kind - `file` or `dir`.

Reply:
> Or file sometimes means item and sometimes path?

Just to clarify: When we want to create new item at remote, we should provide `parent_id` and `title` for new item. In practice this means that we can't create whole path with single request (like `root/dir1/dir2/dir3`). We need to do 3 requests ordered by depth - create dir1, dir2 and dir3.

Reply:
All items on Google Drive can be represented with flat array, where each item has array of parent ids ( item can have multiple parent directories ) and its unique id.

Reply:
> As a related issue we have too many *list* methods.

I've renamed methods, but `list` still in the title. There are 2 recursive generators list methods introduced as implementation of `RemoteBASE.list_cache_paths`. Can't see how it might look better.

Reply:
@Suor resolve?

Reply:
> We have 'item', 'file', 'id', 'path', 'folder'. It is confusing. Let's cut on terminology and preferably stick to whatever we have in other remotes, like files and dirs.

I've unified naming of methods.
`item` is used since PyDrive methods can work both with `file` and `folder` without noticing the difference.